[Python] Test regressions without `-Werror`

#4396 introduced new tests that fail when the test suite is run without `-Werror`, e.g.:

```pytest
$ tox -e py313-cover -- -Wdefault
py313-cover: install_deps> python -I -m pip install -r ../requirements/test.txt
.pkg: _optional_hooks> python /usr/lib/python3.14/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: get_requires_for_build_sdist> python /usr/lib/python3.14/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: build_sdist> python /usr/lib/python3.14/site-packages/pyproject_api/_backend.py True setuptools.build_meta
py313-cover: install_package_deps> python -I -m pip install 'attrs>=22.2.0' 'exceptiongroup>=1.0.0; python_version < "3.11"' 'sortedcontainers<3.0.0,>=2.1.0'
py313-cover: install_package> python -I -m pip install --force-reinstall --no-deps .tox/.tmp/package/4/hypothesis-6.133.2.tar.gz
py313-cover: commands[0]> python -bb -X dev -m pytest -n auto tests/cover/ tests/pytest/ tests/conjecture/ -Wdefault
========================================================= test session starts =========================================================
platform linux -- Python 3.13.3, pytest-8.3.5, pluggy-1.5.0
cachedir: .tox/py313-cover/.pytest_cache
rootdir: /tmp/hypothesis
configfile: pytest.ini
plugins: hypothesis-6.133.2, xdist-3.6.1
12 workers [4208 items]   
............................................................................................................................... [  3%]
............................................................................................................................... [  6%]
............................................................................................................................... [  9%]
............................................................................................................................... [ 12%]
............................................................................................................................... [ 15%]
............................................................................................................................... [ 18%]
............................................................................................................................... [ 21%]
............................................................................................................................... [ 24%]
.................................x.............F.F.F........................................................................... [ 27%]
............................................................................................................................... [ 30%]
...................................................................s........................................................... [ 33%]
............................................................................................................................... [ 36%]
............................................................................................................................... [ 39%]
............................................................................................................................... [ 42%]
.................s...........................................................................................s................. [ 45%]
............................................................................................................................... [ 48%]
............................................................................................................................... [ 51%]
............................................................................................................................... [ 54%]
............................................................................................................................... [ 57%]
............................................................................................................................... [ 60%]
............................................................................................................................... [ 63%]
............................................................................................................................... [ 66%]
............................................................................................................................... [ 69%]
............................................................................................................................... [ 72%]
............................................................................................................................... [ 75%]
............................................................................................................................... [ 78%]
............................................................................................................................... [ 81%]
............................................................................................................................... [ 84%]
............................................................................................................................... [ 87%]
............................................................................................................................... [ 90%]
............................................................................................................................... [ 93%]
...............................................................................................................s............s.. [ 96%]
..x........................................................................s....s.............................................. [ 99%]
.................                                                                                                               [100%]
============================================================== FAILURES ===============================================================
__________________________________ test_compatible_shared_strategies_do_not_raise[strat_a3-strat_b3] __________________________________
[gw8] linux -- Python 3.13.3 /tmp/hypothesis/hypothesis-python/.tox/py313-cover/bin/python
[XPASS(strict)] un-normalized constraint value (issue #4417)
__________________________________ test_compatible_shared_strategies_do_not_raise[strat_a4-strat_b4] __________________________________
[gw8] linux -- Python 3.13.3 /tmp/hypothesis/hypothesis-python/.tox/py313-cover/bin/python
[XPASS(strict)] differently named @composites
__________________________________ test_compatible_shared_strategies_do_not_raise[strat_a5-strat_b5] __________________________________
[gw8] linux -- Python 3.13.3 /tmp/hypothesis/hypothesis-python/.tox/py313-cover/bin/python
[XPASS(strict)] really different (but compatible)
========================================================== warnings summary ===========================================================
hypothesis-python/tests/cover/test_direct_strategies.py::test_compatible_shared_strategies_do_not_raise[strat_a3-strat_b3]
hypothesis-python/tests/cover/test_direct_strategies.py::test_compatible_shared_strategies_do_not_raise[strat_a4-strat_b4]
hypothesis-python/tests/cover/test_direct_strategies.py::test_compatible_shared_strategies_do_not_raise[strat_a5-strat_b5]
  /tmp/hypothesis/hypothesis-python/.tox/py313-cover/lib/python3.13/site-packages/hypothesis/strategies/_internal/shared.py:53: HypothesisWarning: Different strategies are shared under key='share'. This risks drawing values that are not valid examples for the strategy, or that have a narrower range than expected.
    warnings.warn(

hypothesis-python/tests/cover/test_interactive_example.py::test_interactive_example_does_not_emit_warning
  /usr/lib/python3.13/pty.py:95: DeprecationWarning: This process (pid=439193) is multi-threaded, use of forkpty() may lead to deadlocks in the child.
    pid, fd = os.forkpty()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================== slowest 20 durations =========================================================
29.94s call     hypothesis-python/tests/cover/test_slices.py::test_start_will_equal_0
26.13s call     hypothesis-python/tests/cover/test_stateful.py::test_removes_needless_steps
18.13s call     hypothesis-python/tests/conjecture/test_choice.py::test_choice_to_index_injective
12.82s call     hypothesis-python/tests/conjecture/test_engine.py::test_fails_health_check_for_large_non_base
9.95s call     hypothesis-python/tests/conjecture/test_optimiser.py::test_optimiser_when_test_grows_buffer_to_overflow
8.83s call     hypothesis-python/tests/cover/test_lookup.py::test_specialised_mapping_types[typing.ChainMap[tests.cover.test_lookup.Elem, tests.cover.test_lookup.ElemValue]-typing.ChainMap]
7.83s call     hypothesis-python/tests/conjecture/test_choice.py::test_choice_from_value_injective
7.04s call     hypothesis-python/tests/cover/test_statistical_events.py::test_stateful_with_one_of_bundles_states_are_deduped
6.63s call     hypothesis-python/tests/cover/test_database_backend.py::TestGADReads::runTest
6.59s call     hypothesis-python/tests/cover/test_stateful.py::TestMachineUsingMultiple::runTest
6.45s call     hypothesis-python/tests/cover/test_database_backend.py::test_database_listener_memory
6.34s call     hypothesis-python/tests/cover/test_sampled_from.py::test_efficient_lists_of_tuples_first_element_sampled_from
6.26s call     hypothesis-python/tests/cover/test_observability.py::test_observability
5.70s call     hypothesis-python/tests/cover/test_lookup.py::test_specialised_mapping_types[typing.DefaultDict[tests.cover.test_lookup.Elem, tests.cover.test_lookup.ElemValue]-typing.DefaultDict]
5.64s call     hypothesis-python/tests/conjecture/test_engine.py::test_draw_to_overrun
5.43s call     hypothesis-python/tests/cover/test_flakiness.py::test_failure_sequence_inducing
5.43s call     hypothesis-python/tests/cover/test_lookup.py::test_specialised_mapping_types[typing.OrderedDict[tests.cover.test_lookup.Elem, tests.cover.test_lookup.ElemValue]-typing.OrderedDict]
5.41s call     hypothesis-python/tests/cover/test_lookup_py37.py::test_resolving_standard_ordered_dict_as_generic
5.40s call     hypothesis-python/tests/cover/test_lookup_py37.py::test_resolving_standard_itemsview_as_generic
5.36s call     hypothesis-python/tests/cover/test_lookup_py37.py::test_resolving_standard_chainmap_as_generic
======================================================= short test summary info =======================================================
FAILED tests/cover/test_direct_strategies.py::test_compatible_shared_strategies_do_not_raise[strat_a3-strat_b3]
FAILED tests/cover/test_direct_strategies.py::test_compatible_shared_strategies_do_not_raise[strat_a4-strat_b4]
FAILED tests/cover/test_direct_strategies.py::test_compatible_shared_strategies_do_not_raise[strat_a5-strat_b5]
============================= 3 failed, 4196 passed, 7 skipped, 2 xfailed, 4 warnings in 80.03s (0:01:20) =============================
/tmp/hypothesis/hypothesis-python/.tox/py313-cover/lib/python3.13/site-packages/_pytest/pathlib.py:96: PytestWarning: (rm_rf) error removing /tmp/pytest-of-mgorny/garbage-8f79c9c2-4014-4b2c-a365-4ac84bcc3f31/popen-gw3
<class 'OSError'>: [Errno 39] Directory not empty: '/tmp/pytest-of-mgorny/garbage-8f79c9c2-4014-4b2c-a365-4ac84bcc3f31/popen-gw3'
  warnings.warn(
/tmp/hypothesis/hypothesis-python/.tox/py313-cover/lib/python3.13/site-packages/_pytest/pathlib.py:96: PytestWarning: (rm_rf) error removing /tmp/pytest-of-mgorny/garbage-8f79c9c2-4014-4b2c-a365-4ac84bcc3f31
<class 'OSError'>: [Errno 39] Directory not empty: '/tmp/pytest-of-mgorny/garbage-8f79c9c2-4014-4b2c-a365-4ac84bcc3f31'
  warnings.warn(
py313-cover: exit 1 (81.12 seconds) /tmp/hypothesis/hypothesis-python> python -bb -X dev -m pytest -n auto tests/cover/ tests/pytest/ tests/conjecture/ -Wdefault pid=439159
  py313-cover: FAIL code 1 (87.08=setup[5.95]+cmd[81.12] seconds)
  evaluation failed :( (87.21 seconds)
```

<del>The tests are actually doubly wrong:</del>
- <del>they use `pytest.mark.xfail(strict=True...` instead of `pytest.raises()` to assert for exceptions</del>
- <del>they rely on `-Werror` instead of using `pytest.warns()` to assert for warnings</del>

CC @jobh

--- a/hypothesis-python/tests/cover/test_direct_strategies.py
+++ b/hypothesis-python/tests/cover/test_direct_strategies.py
@@ -13,6 +13,7 @@ import decimal
 import enum
 import fractions
 import math
+import warnings
 from datetime import date, datetime, time, timedelta
 from ipaddress import IPv4Network, IPv6Network
 
@@ -671,4 +672,6 @@ def test_compatible_shared_strategies_do_not_raise(strat_a, strat_b):
     def test_it(a, b):
         assert a == b
 
-    test_it()
+    with warnings.catch_warnings():
+        warnings.simplefilter("error", HypothesisWarning)
+        test_it()

